# Assignment 2 Rough Work

In [2]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC, SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, log_loss
from pandas.api.types import CategoricalDtype
%matplotlib inline

## 1. Exploratory Analysis

In [32]:
# create data frames
data_path = os.path.join(os.getcwd(), 'Data', 'PartA', '20ng_train.csv')
trn_newsgroup = pd.read_csv(data_path, delimiter=',', compression='bz2')
data_path2 = os.path.join(os.getcwd(), 'Data', 'PartA', '20ng_test.csv')
tst_newsgroup = pd.read_csv(data_path2, delimiter=',', compression='bz2')
data_path3 = os.path.join(os.getcwd(), 'Data', 'PartA', '20ng_labels.csv')
lbl_newsgroup = pd.read_csv(data_path3, delimiter=',', compression='bz2')

0
0  comp.sys.ibm.pc.hardware
1     comp.sys.mac.hardware
2                 rec.autos
3           rec.motorcycles
4                 sci.crypt
5           sci.electronics
6    soc.religion.christian
7        talk.religion.misc

In [22]:
# drop class labels
trn_newsgroup_targets = trn_newsgroup['class']
trn_newsgroup = trn_newsgroup.drop(['class'], axis = 1)
tst_newsgroup_targets = tst_newsgroup['class']
tst_newsgroup = tst_newsgroup.drop(['class'], axis = 1)

In [54]:
print('Number of train rows: {}'.format(trn_newsgroup.shape[0]))
print('Number of train columns: {}'.format(trn_newsgroup.shape[1]))
print('Number of test rows: {}'.format(tst_newsgroup.shape[0]))
print('Number of test columns: {}'.format(tst_newsgroup.shape[1]))
trn_newsgroup.describe()
tst_newsgroup.describe()

zeros = 0
ones = 0 
twos = 0
threes = 0
fours = 0
fives = 0
sixes = 0
sevens = 0
for row in trn_newsgroup_targets:
    if(trn_newsgroup_targets[row] == 0): 
        zeros = zeros + 1 
    if(trn_newsgroup_targets[row] == 1): 
        ones = ones + 1
    if(trn_newsgroup_targets[row] == 2): 
        twos = twos + 1
    if(trn_newsgroup_targets[row] == 3): 
        threes = threes + 1
    if(trn_newsgroup_targets[row] == 4): 
        fours = fours + 1
    if(trn_newsgroup_targets[row] == 5): 
        fives = fives + 1
    if(trn_newsgroup_targets[row] == 6): 
        sixes = sixes + 1
    if(trn_newsgroup_targets[row] == 7): 
        sevens = sevens + 1
print(zeros)
print(ones)
print(twos)
print(threes)
print(fours)
print(fives)
print(sixes)
print(sevens)
print(zeros+ones+twos+threes+fours+fives+sixes+sevens)

Number of train rows: 5648
Number of train columns: 1001
Number of test rows: 1883
Number of test columns: 1001
0
1489
1193
737
1486
0
0
743
5648


### ==== Question 1.1 =====

Looking at the training set, we see that there are 5648 entries of which 1000 words and their frequencies are being checked. Obviously when dealing with large texts the word frequencies will often be 0 or extremely low. This is the case in our data set as the mean is 0.002292. There are 8 possible classes that the classifier can label an entry as. And from looking at the distribution of these 

### ==== Question 1.2 ====

The testing has 1883 entries of newsgroup documents. Resulting in a 75 to 25 percent split in training to testing data. This decision to have significantly more training entries brings its drawbacks and benefits. The larger you make a training set the more information the classifier will have to improve its accuracy during testing and in practice. However, the more testing data you have, the more we understand how accurate the classifier will be in practice. Essentially in prioritising training, we are gaining a more accurate predictor while losing our ability to know how  accurate the classifier will be on previously unseen entries. 

### ==== Question 1.3 ====